docs/codeql/codeql-language-guides/customizing-library-models-for-go.rst

```diff
@@ -0,0 +1,319 @@
+- The fourth value ``Hostname`` is the method name.
+- The fifth value ``""`` is left blank, since specifying the signature is optional and Go does not allow multiple signature overloads for the same function.
+
+The sixth value should be left empty and is out of scope for this documentation.
+The remaining values are used to define the ``access path``, the ``kind``, and the ``provenance`` (origin) of the summary.
+
+- The seventh value is the access path to the input (where data flows from). ``Argument[receiver]`` is the access path to the receiver (``u`` in the example).
+- The eighth value ``ReturnValue`` is the access path to the output (where data flows to), in this case ``ReturnValue``, which means that the input flows to the return value.
+- The ninth value ``taint`` is the kind of the flow. ``taint`` means that taint is propagated through the call.
+- The tenth value ``manual`` is the provenance of the summary, which is used to identify the origin of the summary.
+
+Example: Accessing the ``Body`` field of an HTTP request
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This example shows how we can model a field read as a source of tainted data.
+
+.. code-block:: go
+
+   func TaintFlow(w http.ResponseWriter, r *http.Request) {
+       body := r.Body // The Body field of an HTTP request is a source of tainted data.
+       ...
+   }
+
+We need to add a tuple to the ``sourceModel``\(package, type, subtypes, name, signature, ext, output, kind, provenance) extensible predicate by updating a data extension file.
+
+.. code-block:: yaml
+
+  extensions:
+  - addsTo:
+      pack: codeql/go-all
+      extensible: sourceModel
+    data:
+      - ["net/http", "Request", True, "Body", "", "", "", "remote", "manual"]
+
+Since we are adding a new source, we need to add a tuple to the ``sourceModel`` extensible predicate.
+The first five values identify the field to be modeled as a source.
+
+- The first value ``net/http`` is the package name.
+- The second value ``Request`` is the name of the type that the field is associated with.
+- The third value ``True`` is a flag that indicates whether or not the source also applies to all overrides of the field.
+- The fourth value ``Body`` is the field name.
+- The fifth value ``""`` is blank since it is a field access and field accesses do not have method signatures in Go.
+
+The sixth value should be left empty and is out of scope for this documentation.
+The remaining values are used to define the ``access path``, the ``kind``, and the ``provenance`` (origin) of the source.
+
+- The seventh value ``""`` is left blank. Leaving the access path of a source model blank indicates that it is a field access.
+- The eighth value ``remote`` is the source kind. This indicates that the source is a remote source of untrusted data.
+- The ninth value ``manual`` is the provenance of the source, which is used to identify the origin of the source.
+
```

Comment:
```suggestion

Package versions
~~~~~~~~~~~~~~~~

For versions greater than 1, the version number is included in the package import path, usually as ``/v2``. This is called the major version suffix. We normally want our models to apply to all versions of a package. Rather than having to repeat models with the package column changed to include all available versions, we can just use the package name without the major version suffix and this will be matched to any version. So models with ``github.com/couchbase/gocb`` in the package column will match packages imported from ``github.com/couchbase/gocb`` and ``github.com/couchbase/gocb/v2`` (or any other version). Note that packages hosted at ``gopkg.in`` use a slightly different syntax: the major version suffix looks like ``.v2`` and is present even for version 1. This is also supported. So models with ``gopkg.in/yaml`` in the package column will match packages imported from ``gopkg.in/yaml.v1``, ``gopkg.in/yaml.v2`` and ``gopkg.in/yaml.v3``.

To write models that only apply to ``github.com/couchbase/gocb/v2``, it is sufficient to include the major version suffix (``/v2``) in the package column. To write models that only apply to ``github.com/couchbase/gocb``, you may prefix the package column with ``fixed-version:``. For example, here are two models for a method that has changed name from v1 to v2.

.. code-block:: yaml
  extensions:
  - addsTo:
      pack: codeql/go-all
      extensible: sinkModel
    data:
      - ["fixed-version:github.com/couchbase/gocb", "Cluster", True, "ExecuteAnalyticsQuery", "", "", "Argument[0]", "nosql-injection", "manual"]
      - ["github.com/couchbase/gocb/v2", "Cluster", True, "AnalyticsQuery", "", "", "Argument[0]", "nosql-injection", "manual"]

```